src/map/script.cpp

```diff
@@ -25686,6 +25722,281 @@ BUILDIN_FUNC( openstylist ){
 #endif
 }
 
+BUILDIN_FUNC(getitemdata)
+{
+	struct map_session_data* sd;
+
+	if (!script_rid2sd(sd)) {
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	int idx = script_getnum(st, 2);
+	if (idx < 0 || idx >= MAX_INVENTORY) {
+		ShowError("buildin_getitemdata: Index %d is out of the range 0-%d.\n", idx, MAX_INVENTORY - 1);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	if (sd->inventory_data[idx] == nullptr) {
+		ShowError("buildin_getitemdata: No item found on index %d of player %s (AID: %u, CID: %u).\n", idx, sd->status.name, sd->status.account_id, sd->status.char_id);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+	if (!itemdb_isequip2(sd->inventory_data[idx]) || sd->inventory_data[idx]->type == IT_AMMO) {
+		ShowError("buildin_getitemdata: Invalid type for item %u. The item must be equippable.\n", sd->inventory_data[idx]->nameid);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	struct script_data *data = script_getdata(st, 3);
+	char* name = reference_getname(data);
+
+#define getitemdata_sub(idx__,var__) setd_sub_num(st,sd,name,(idx__),(var__),data->ref)
+
+	struct item *item_tmp = &sd->inventory.u.items_inventory[idx];
+	uint8 i;
+
+	getitemdata_sub( SID_ID, item_tmp->nameid );
+	getitemdata_sub( SID_IDENTIFY, item_tmp->identify );
+	getitemdata_sub( SID_REFINE, item_tmp->refine );
+	getitemdata_sub( SID_ATTRIBUTE, item_tmp->attribute );
+	for (i = 0; i < MAX_SLOTS; i++)
+		getitemdata_sub( SID_CARD1 + i, item_tmp->card[i] );
+	getitemdata_sub( SID_EXPIRE, item_tmp->expire_time );
+	getitemdata_sub( SID_BOUND, item_tmp->bound );
+	getitemdata_sub( SID_GRADE, item_tmp->enchantgrade );
+	for (i = 0; i < MAX_ITEM_RDM_OPT; i++) {
+		getitemdata_sub( SID_OPTIONID1 + i, item_tmp->option[i].id );
+		getitemdata_sub( SID_OPTIONVALUE1 + i, item_tmp->option[i].value );
+		getitemdata_sub( SID_OPTIONPARAM1 + i, item_tmp->option[i].param );
+	}
+
+#undef getitemdata_sub
+
+	return SCRIPT_CMD_SUCCESS;
+}
+
+BUILDIN_FUNC(setitemdata)
+{
+	map_session_data* sd;
+
+	if (!script_rid2sd(sd)) {
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	int idx = script_getnum(st, 2);
+	if (idx < 0 || idx >= MAX_INVENTORY) {
+		ShowError("buildin_setitemdata: Index %d is out of the range 0-%d.\n", idx, MAX_INVENTORY - 1);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	if (sd->inventory_data[idx] == nullptr) {
+		ShowError("buildin_setitemdata: No item found on index %d of player %s (AID: %u, CID: %u).\n", idx, sd->status.name, sd->status.account_id, sd->status.char_id);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+	if (!itemdb_isequip2(sd->inventory_data[idx]) || sd->inventory_data[idx]->type == IT_AMMO) {
+		ShowError("buildin_setitemdata: Invalid type for item %u. The item must be equippable.\n", sd->inventory_data[idx]->nameid);
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+
```

Comment:
Just a thought for the "set" function:
Should we block calls if the target item is equipped or in equip switch?
(If we allow during equip: should we run status calculations or whatever after changing data?)